gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java

```diff
@@ -267,4 +209,119 @@ private Instant getCutoffTimeForRef(String reference, Map<String, Instant> dropp
                 GCUtil.deserializeReference(reference).getName(),
                 gcParams.getDefaultCutOffTimestamp());
   }
+
+  // --- Methods for computing expired content rows ----------
+  /**
+   * convert scala Iterator of reference into scala Iterator of {@link Row}.
+   *
+   * <p>Each reference can produce one or more rows.
+   */
+  private scala.collection.Iterator<Row> getExpiredContentRows(
+      scala.collection.Iterator<String> references,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap,
+      String runId,
+      Timestamp startedAt) {
+    List<Iterator<Row>> iterators = new ArrayList<>();
+    try (NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs())) {
```

Comment:
You shouldn't use the try-with-resources.
The spark transformations are lazy and async, and the try-with-resources won't know when it should close the Nessie API.

you can try something like this:
On the driver, create a `Broadcast` variable of `NessieApiV1`
It is a way of passing states in Sparks such as clients between driver and executors.
You should not close it

Then pass the `Broadcast<NessieAPI>` into your transformation
it will create N instances of nessieAPI, one per executor.
After all, the processing is done (and after saving to a table) you need to close the Nessie API explicitly.

Reply:
how it brodcasts? I assuming by serilaizing it and some inside content of `NessieApiV1` has Http stuff, which are not serializable.

> Alternatively, you can keep creating API on the executor as is now, but you need to take care of closing it manually once all the processing is done.

I just return iterator of rows to driver and so spark it self will process the iterator, call doesn't comes back to executor to close it.

Reply:
Tried broadcast, broadcast goes for serialization.

I had to make 10 plus classes serializable for http client, even after that `com.fasterxml.jackson.databind.jsontype.impl.TypeNameIdResolver` is not serializable